sshd-common/src/main/java/org/apache/sshd/common/util/buffer/Buffer.java

```diff
@@ -348,29 +348,36 @@ public List<OpenSshCertificate.CertificateOption> getCertificateOptions() {
 
     /**
      * According to <A HREF=
-     * "https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.certkeys#L222-L262">PROTOCOL.certkeys</A>:
+     * "https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.certkeys#L221-L319">PROTOCOL.certkeys</A>:
      *
      * Critical Options is a set of bytes that is
      *
-     * [overall length][name(string)][data(string)]...
+     * [overall length][name(string)][[length of buffer][[length of string][data(string)]]]...
      *
-     * Where each Critical Option is encoded as a name (string) and data (string)
+     * Where each Certificate Option is encoded as a name (string) and buffer of data (string packed in a buffer)
      *
-     * Then the entire name + data strings are added as bytes (which will get a length prefix)
+     * Then the entire name (string) + data (buffer) are added as bytes (which will get a length prefix)
      *
      * @param  charset {@link Charset} to use for converting bytes to characters
      * @return         the parsed result, never {@code null}, but possibly empty
      */
     public List<OpenSshCertificate.CertificateOption> getCertificateOptions(Charset charset) {
-        // pull out entire Critical Options section
-        final ByteArrayBuffer optionBuffer = new ByteArrayBuffer(getBytes());
-
         List<OpenSshCertificate.CertificateOption> list = new ArrayList<>();
 
-        while (optionBuffer.available() > 0) {
-            String name = optionBuffer.getString(charset);
-            String data = GenericUtils.trimToEmpty(optionBuffer.getString(charset));
-            list.add(new OpenSshCertificate.CertificateOption(name, data.length() > 0 ? data : null));
+        if (available() > 0) {
+            // pull out entire Certificate Options section
+            final ByteArrayBuffer optionBuffer = new ByteArrayBuffer(getBytes());
```

Comment:
Why can't it be `Buffer optionBuffer = new ...` - are you using any *ByteArrayBuffer* methods that are not available via the *Buffer* class ?

Hint: recommend to use the "weakest" variable type you need and not what you know its full type is.

Reply:
@lgoldstein same as previous comment